Allow to specify range specification instead of fixed version

potiuk opened the pull request:
That allows to install for example the latest bug-fix version of
terraform 1.12.* even if 1.13 is already installed.

potiuk commented:
I'd love to get this one merged :)

potiuk commented:
Any hopes of attention from maintainers please :)?

mik-laj commented:
@sudomateo  @chenrui333  Can you help me with it? I really want to use it.

alexjurkiewicz commented:
This PR should support version strings with tilde prefix. `~0.12.0` would support all `0.12.x` versions and is IMO clearer than `<0.13.0`.

potiuk commented:
Sure. Good point. I can add this possibility. I am using semver library and it supports ~ so I can extend it to support strings starting with any of the supported matchers.

potiuk commented:
All right - this is now much nicer - all ranges supported by semver are also supported, I've added tests covering several cases  of ranges matching (and I added checks if matched version is as expected).

potiuk commented:
I also corrected some grammar while updating the docs for it :)

alexjurkiewicz commented:
That's wonderful stuff. It looks like you have given up waiting for Hashicorp and are publishing your own fork? Are you planning to maintain long-term? 🤞

potiuk commented:
Well. I have not given up (yet, dear Hashicorp maintainers), I just published it for my own use as I needed it it in my own CI pipeline :).

I am happy to release a new version in my own fork for now though. Give me a minute.

potiuk commented:
There you are @alexjurkiewicz https://github.com/marketplace/actions/hashicorp-setup-terraform-with-constraints\

@sudomateo @chenrui333 -> I'd really love that one to be merged so that people won't have to rely on my fork :).

If there is anything else I can do to get it merged, please let me know.

sudomateo commented on README.md:
```diff
@@ -136,19 +136,28 @@ steps:
+   place within the credentials block of the Terraform CLI configuration file.
 
-- `terraform_version` - (optional) The version of Terraform CLI to install. A value of `latest` will install the latest version of Terraform CLI. Defaults to `latest`.
+- `terraform_version` - (optional) The version of Terraform CLI to install. Instead of full version string you
```
```suggestion
- `terraform_version` - (optional) The version of Terraform CLI to install. Instead of a full version string you
```

sudomateo commented on README.md:
```diff
@@ -136,19 +136,28 @@ steps:
 
-- `terraform_version` - (optional) The version of Terraform CLI to install. A value of `latest` will install the latest version of Terraform CLI. Defaults to `latest`.
+- `terraform_version` - (optional) The version of Terraform CLI to install. Instead of full version string you
+   can also specify constraint string (see [Semver Ranges](https://www.npmjs.com/package/semver#ranges)
```
```suggestion
   can also specify a constraint string (see [Semver Ranges](https://www.npmjs.com/package/semver#ranges)
```

sudomateo commented on action.yml:
```diff
@@ -1,6 +1,6 @@
-name: 'HashiCorp - Setup Terraform'
+name: 'HashiCorp - Setup Terraform With Constraints'
```
```suggestion
name: 'HashiCorp - Setup Terraform'
```

sudomateo commented on action.yml:
```diff
@@ -1,6 +1,6 @@
+name: 'HashiCorp - Setup Terraform With Constraints'
 description: 'Sets up Terraform CLI in your GitHub Actions workflow.'
-author: 'HashiCorp, Inc.'
+author: 'HashiCorp, Inc. (original) with jarek@potiuk.com modifications'
```
```suggestion
author: 'HashiCorp, Inc.'
```

sudomateo reviewed (changes requested):
Thank you for the work on this. I was on vacation for the last few weeks. I'm also not the maintainer of this project anymore. Since I was pinged a few times I took a look at it.

potiuk commented:
> Thank you for the work on this. I was on vacation for the last few weeks. I'm also not the maintainer of this project anymore. Since I was pinged a few times I took a look at it.

Thanks! I understand, and sorry for the pinging, but it's hard to know whom can we ping.
All your comments are addressed now.

sudomateo commented on action.yml:
```diff
@@ -1,6 +1,6 @@
 name: 'HashiCorp - Setup Terraform'
 description: 'Sets up Terraform CLI in your GitHub Actions workflow.'
-author: 'HashiCorp, Inc.'
+author: 'HashiCorp, Inc. (original)'
```
```suggestion
author: 'HashiCorp, Inc.'
```

potiuk replied:
Sorry - resolved that one too :).

sudomateo reviewed (changes requested):
I hear you. I'm still trying to find a maintainer for this project going forward, but it's proving to be difficult thanks to COVID-19.

potiuk commented:
> I hear you. I'm still trying to find a maintainer for this project going forward, but it's proving to be difficult thanks to COVID-19.

Hmm. Happy to take it over, but I guess this cannot be done if I am not a Hashicorp employee.

sudomateo commented:
@potiuk I can always ask. I believe we'd need to move this to the `hashicorp-community` organization to do that.

Either way, I began testing these changes in https://github.com/sudomateo/urban-bassoon/actions and I was not able to get the `~0.12` or `0.12.x` syntax working. Only `<0.13.0` worked.

potiuk commented:
> @potiuk I can always ask. I believe we'd need to move this to the `hashicorp-community` organization to do that.
> 
> Either way, I began testing these changes in https://github.com/sudomateo/urban-bassoon/actions and I was not able to get the `~0.12` or `0.12.x` syntax working. Only `<0.13.0` worked.

> Run potiuk/setup-terraform@master

You used run potiuk/setup-terraform@master and I released it in my fork with v1.3.0 version

potiuk commented:
This is the weird thing about Github Actions - the tip of branches have nothing to do with marketplace releases. I can release master now :)

potiuk commented:
Ach no - it would work normally with master as well, but since I changed the name back, it is not automatically moving the tag becuase the name is duplicate of the Hashicorp one :woman_shrugging:

potiuk commented:
v1.3.0 is still with the modified name so you can use it for testing. The only difference are the corrected descriptions

potiuk commented:
![Screenshot from 2020-09-08 15-44-28](https://user-images.githubusercontent.com/595491/92484652-40cc6600-f1ea-11ea-83ae-697fc4c8858e.png)

sudomateo commented:
Testing with `v1.3.0` now.

potiuk commented:
Wohoo!

mik-laj commented:
:tada:

@sudomateo @potiuk Thank you.

potiuk commented:
Thanks @sudomateo -> let me know about the maintenance thing. It's not a lot of work, this action is rather useful, and since I leave and breathe open-source for 2 years now, I am happy to do some of those things outside of the main community I contribute to (Apache Airflow). If you'd plan to make it full community driven maybe we can have a small group of people to volunteer to maintain this an possibly few others tools around open-source Hashicorp stuff. I think Hashicorp had contributed a lot to the community and giving back is a good thing to do.

We've added Hashicorp Vault Secret Backend recently (I extended it and added native Hook integration to Vault) and very recently we started to discuss native Terraform integration for Apache Airflow  https://github.com/apache/airflow/issues/10454 and I know at least few people who love Terraform's tools and with open-source mindset so maybe we can help somehow.

mik-laj commented:
We also working on a terraform provider for Apache Airflow: https://github.com/apache/airflow/issues/10550

potiuk commented:
Indeed @mik-laj :). 

I de-listed my forked action from the marketplace now BTW.

alexjurkiewicz commented:
Thank you @potiuk and @sudomateo ❤️
I would like to put my hand up for helping maintain this too.
